File: src/main/services/PluginGuid.ts
import fs from 'fs';
import path from 'path';
const TOKEN_RE = /[A-Za-z][A-Za-z0-9_]{1,40}(?:\.[A-Za-z][A-Za-z0-9_]{0,40}){2,6}/g;
const UTF16_RUN_RE = /(?:[A-Za-z0-9_.]\x00){3,}/g;
const STOP_SEGMENTS = new Set([
    'system',
    'unity',
    'unityengine',
    'microsoft',
    'mscorlib',
    'mono',
    'monomod',
    'newtonsoft',
    'bepinex',
    'csharp',
    'assembly',
    'js',
    'native',
    'dotnet'
]);
const STOP_TAIL = /\.(patches|ui|optionsapi|assettools|configuration|logging|bootstrap|api|tool|editor|handler|objectpool|extensions|manager|debug|state|core)$/i;
const STOP_MID = new Set([
    'ui',
    'editor',
    'tools',
    'extensions',
    'handlers',
    'objectpool',
    'manager',
    'debug',
    'state',
    'core',
    'patches',
    'api'
]);
function scanTokens(text: string): {
    utf16: string[];
    ascii: string[];
} {
    const ascii = text.match(TOKEN_RE) || [];
    const utf16 = (text.match(UTF16_RUN_RE) || []).map((run) => run.replace(/\x00/g, ''));
    const plausible = (raw: string[]): string[] => {
        const seen = new Set<string>();
        const out: string[] = [];
        for (const rawTok of raw) {
            const segs = rawTok.split('.').filter(Boolean);
            const len = segs.length;
            if (len < 3 || len > 7)
                continue;
            if (segs.some((s) => STOP_SEGMENTS.has(s.toLowerCase())))
                continue;
            if (segs.some((s) => /^\d+$/.test(s)))
                continue;
            const tok = segs.join('.');
            if (seen.has(tok))
                continue;
            if (STOP_TAIL.test(tok))
                continue;
            const mid = segs.slice(1, -1).map((s) => s.toLowerCase());
            if (mid.some((s) => STOP_MID.has(s)))
                continue;
            seen.add(tok);
            out.push(tok);
        }
        return out;
    };
    const utf = plausible(utf16).sort((a, b) => a.split('.').length - b.split('.').length);
    const asciiOut = plausible(ascii)
        .sort((a, b) => a.split('.').length - b.split('.').length)
        .filter((a) => !utf.includes(a));
    return { utf16: utf, ascii: asciiOut };
}
const CANDIDATE_CACHE = new Map<string, string[]>();
const CANDIDATE_CACHE_MAX = 1024;
function cacheKeyFor(dllPath: string): string | null {
    try {
        const st = fs.statSync(dllPath);
        if (!st.isFile())
            return null;
        const dir = path.dirname(dllPath).toLowerCase();
        return `${dir}\u0000${path.basename(dllPath).toLowerCase()}\u0000${st.size}\u0000${st.mtimeMs}`;
    }
    catch {
        return null;
    }
}
function computeCandidates(dllPath: string): string[] {
    let text: string;
    try {
        text = fs.readFileSync(dllPath).toString('latin1');
    }
    catch {
        return [];
    }
    const { utf16, ascii } = scanTokens(text);
    return [...utf16, ...ascii];
}
export function extractPluginCandidates(dllPath: string): string[] {
    if (!fs.existsSync(dllPath))
        return [];
    const key = cacheKeyFor(dllPath);
    if (!key)
        return [];
    const cached = CANDIDATE_CACHE.get(key);
    if (cached)
        return cached.slice();
    const value = computeCandidates(dllPath);
    if (CANDIDATE_CACHE.size >= CANDIDATE_CACHE_MAX) {
        const oldest = CANDIDATE_CACHE.keys().next().value;
        if (oldest !== undefined)
            CANDIDATE_CACHE.delete(oldest);
    }
    CANDIDATE_CACHE.set(key, value);
    return value.slice();
}
export function extractPluginGuid(dllPath: string): string | undefined {
    return extractPluginCandidates(dllPath)[0];
}

File: RBBPMM.Wpf/RBBPMM.Core/PluginGuid.cs
using System.Text;
using System.Text.RegularExpressions;

namespace RBBPMM.Core;

/// <summary>
/// Extracts BepInEx plugin GUID candidates straight out of a compiled <c>.dll</c> by scanning its
/// string table — no assembly loading, so a malformed or hostile plugin can never execute here.
/// Ported from src/main/services/PluginGuid.ts.
/// </summary>
public static class PluginGuid
{
    // Dotted identifiers with 2..6 extra segments, e.g. "Some.Author.CoolMod".
    private static readonly Regex TokenRe = new(
        @"[A-Za-z][A-Za-z0-9_]{1,40}(?:\.[A-Za-z][A-Za-z0-9_]{0,40}){2,6}",
        RegexOptions.Compiled);

    // UTF-16LE runs appear as "X\0Y\0Z\0" once the file is decoded as Latin-1.
    private static readonly Regex Utf16RunRe = new(
        @"(?:[A-Za-z0-9_.]\x00){3,}",
        RegexOptions.Compiled);

    private static readonly Regex StopTail = new(
        @"\.(patches|ui|optionsapi|assettools|configuration|logging|bootstrap|api|tool|editor|handler|objectpool|extensions|manager|debug|state|core)$",
        RegexOptions.Compiled | RegexOptions.IgnoreCase);

    private static readonly Regex NumericSegment = new(@"^\d+$", RegexOptions.Compiled);

    private static readonly HashSet<string> StopSegments = new(StringComparer.OrdinalIgnoreCase)
    {
        "system", "unity", "unityengine", "microsoft", "mscorlib", "mono", "monomod",
        "newtonsoft", "bepinex", "csharp", "assembly", "js", "native", "dotnet"
    };

    private static readonly HashSet<string> StopMid = new(StringComparer.OrdinalIgnoreCase)
    {
        "ui", "editor", "tools", "extensions", "handlers", "objectpool", "manager",
        "debug", "state", "core", "patches", "api"
    };

    private const int CandidateCacheMax = 1024;
    private static readonly Dictionary<string, string[]> CandidateCache = new();
    private static readonly Lock CacheLock = new();

    public static (List<string> Utf16, List<string> Ascii) ScanTokens(string text)
    {
        var ascii = TokenRe.Matches(text).Select(m => m.Value);
        var utf16 = Utf16RunRe.Matches(text).Select(m => m.Value.Replace("\0", ""));

        var utf = Plausible(utf16).OrderBy(t => t.Count(c => c == '.')).ToList();
        var asciiOut = Plausible(ascii)
            .OrderBy(t => t.Count(c => c == '.'))
            .Where(t => !utf.Contains(t))
            .ToList();

        return (utf, asciiOut);
    }

    private static List<string> Plausible(IEnumerable<string> raw)
    {
        var seen = new HashSet<string>(StringComparer.Ordinal);
        var outList = new List<string>();

        foreach (var rawTok in raw)
        {
            var segs = rawTok.Split('.', StringSplitOptions.RemoveEmptyEntries);
            if (segs.Length < 3 || segs.Length > 7)
                continue;
            if (segs.Any(s => StopSegments.Contains(s)))
                continue;
            if (segs.Any(s => NumericSegment.IsMatch(s)))
                continue;

            var tok = string.Join('.', segs);
            if (!seen.Add(tok))
                continue;
            if (StopTail.IsMatch(tok))
                continue;

            var mid = segs.Skip(1).Take(segs.Length - 2).ToList();
            if (mid.Any(s => StopMid.Contains(s)))
                continue;

            outList.Add(tok);
        }

        return outList;
    }

    /// <summary>Key on directory + name + size + mtime so a rewritten dll invalidates its entry.</summary>
    private static string? CacheKeyFor(string dllPath)
    {
        try
        {
            var fi = new FileInfo(dllPath);
            if (!fi.Exists)
                return null;
            return $"{Path.GetDirectoryName(dllPath)?.ToLowerInvariant()}\0{fi.Name.ToLowerInvariant()}\0{fi.Length}\0{fi.LastWriteTimeUtc.Ticks}";
        }
        catch
        {
            return null;
        }
    }

    private static List<string> ComputeCandidates(string dllPath)
    {
        string text;
        try
        {
            text = Encoding.Latin1.GetString(File.ReadAllBytes(dllPath));
        }
        catch
        {
            return [];
        }

        var (utf16, ascii) = ScanTokens(text);
        return [.. utf16, .. ascii];
    }

    /// <summary>Ordered GUID candidates for a plugin dll (UTF-16 hits first, then ASCII-only).</summary>
    public static List<string> ExtractPluginCandidates(string dllPath)
    {
        if (!File.Exists(dllPath))
            return [];

        var key = CacheKeyFor(dllPath);
        if (key == null)
            return [];

        lock (CacheLock)
        {
            if (CandidateCache.TryGetValue(key, out var cached))
                return [.. cached];

            var value = ComputeCandidates(dllPath);

            if (CandidateCache.Count >= CandidateCacheMax)
            {
                var oldest = CandidateCache.Keys.FirstOrDefault();
                if (oldest != null)
                    CandidateCache.Remove(oldest);
            }

            CandidateCache[key] = [.. value];
            return value;
        }
    }

    /// <summary>Best-effort plugin GUID — the first candidate, or <c>null</c>.</summary>
    public static string? ExtractPluginGuid(string dllPath)
    {
        var candidates = ExtractPluginCandidates(dllPath);
        return candidates.Count > 0 ? candidates[0] : null;
    }
}
